entropylab/api/param_store.py

```diff
@@ -0,0 +1,322 @@
+
+    """ Items """
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        self._params[key] = value
+        self._is_dirty = True
+
+    def __getitem__(self, key: str) -> Any:
+        return self._params[key]
+
+    def __delitem__(self, *args, **kwargs):
+        self._params.__delitem__(*args, **kwargs)
+        self._is_dirty = True
+
+    def __contains__(self, key: str) -> bool:
+        return key in self._params
+
+    def to_dict(self) -> Dict:
+        return dict(self._params)
+
+    def get(self, key: str, commit_id: Optional[str] = None):
+        """
+            returns the value of a param by key
+
+        :param key: the key identifying the param
+        :param commit_id: an optional commit_id. if provided, the value will be
+        returned from the specified commit
+        """
+        if commit_id is None:
+            return self._params.get(key)
+        else:
+            commit = self._get_commit(commit_id)
+            return commit["params"][key]
+
+    """ Commits """
+
+    def commit(self, label: Optional[str] = None) -> str:
+        if not self._is_dirty:
+            return self._base_commit_id
+        metadata = self._generate_metadata(label)
+        if self._is_in_memory_mode:
+            params = self._deep_copy(self._params)
+        else:
+            params = self._params
+        doc_id = self._db.insert(dict(metadata=metadata.__dict__, params=params))
+        self._base_commit_id = metadata.id
+        self._base_doc_id = doc_id
+        self._is_dirty = False
+        return metadata.id
+
+    def checkout(
+        self,
+        commit_id: Optional[str] = None,
+        commit_num: Optional[int] = None,
+        move_by: Optional[int] = None,
+    ) -> None:
+        commit = self._get_commit(commit_id, commit_num, move_by)
+        self._params = commit["params"]
+        self._base_commit_id = commit_id
+        self._base_doc_id = commit.doc_id
+        self._is_dirty = False
+
+    def list_commits(self, label: Optional[str] = None) -> List[Metadata]:
+        """
+            returns a list of commits
+
+        :param label: an optional label, if given then only commits that match
+        it will be returned
+        """
+        documents = self._db.search(
+            Query().metadata.label.test(_test_if_value_contains(label))
+        )
+        metadata = map(_extract_metadata, documents)
+        return list(metadata)
+
+    def _generate_metadata(self, label: Optional[str] = None) -> (str, int):
+        metadata = Metadata()
+        metadata.ns = time.time_ns()
+        params_json = json.dumps(self._params, sort_keys=True, ensure_ascii=True)
+        commit_encoded = (params_json + str(metadata.ns)).encode("utf-8")
+        metadata.id = hashlib.sha1(commit_encoded).hexdigest()
+        metadata.label = label
+        return metadata
+
+    @staticmethod
+    def _deep_copy(d: Dict) -> Dict:
+        return json.loads(json.dumps(d))
```

Comment:
why do we need the json lib here?

Reply:
It's a simple way to deep copy a dictionary.

Reply:
do we enforce everything to be json compatible? i guess we are, just need to validate no different error messages will arise here..